src/internal/log/span.go

```diff
@@ -0,0 +1,138 @@
+package log
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// Level is the level at which to generate Span logs.
+type Level int
+
+const (
+	DebugLevel Level = 1
+	InfoLevel  Level = 2
+	ErrorLevel Level = 3
+)
+
+func (l Level) log(z *zap.Logger, msg string, fields ...Field) {
+	switch l { //exhaustive:enforce
+	case DebugLevel:
+		z.Debug(msg, fields...)
+		return
+	case InfoLevel:
+		z.Info(msg, fields...)
+		return
+	case ErrorLevel:
+		z.Error(msg, fields...)
+		return
+	}
+	z.DPanic("log: internal error: unknown level in Level.log call", zap.Int("level", int(l)))
+	z.Debug(msg, fields...)
+}
+
+// EndSpanFunc is a function that ends a span.
+type EndSpanFunc = func(fields ...Field)
+
+// ErrorL is a Field that marks a span as failed, and logs the end of the span at the provided level.
+func ErrorL(err error, level Level) Field {
+	if err == nil {
+		return zap.Skip()
+	}
+	f := zap.Error(err)
+	f.Integer = int64(level)
+	return f
+}
+
+// ErrorpL is a Field that marks a span as failed if *err is a non-nil error at the time when the
+// span ends.  See ErrorL.
+func ErrorpL(err *error, level Level) Field {
+	// If you're wondering why this ends in "pL" with that capitalization; zap uses the
+	// convention 'lowercase p = pointer', and we use the convention 'uppercase L = level'.
+	f := zap.Reflect("error", err)
```

Comment:
FWIW, I noticed that the [`zap` docs for `Reflect`](https://pkg.go.dev/go.uber.org/zap#Reflect) state ‘outside tests, `Any` is always a better choice *[than `Reflect`]*.’

Reply:
Basically Reflect decides the type at log message marshaling time, and Any decides at the time you call Any.  We sneak in there and change the internals of the field before this is passed anywhere.  It might be clearest if we `zapcore.Field{Type: ErrorField, Interface: err}` or something.

Reply:
I actually ended up making this its own field type, so there should be no confusion now.  The implementation is a little bit hack (we just choose a number much larger than any existing zapcore field type), but it probably won't be a problem.